### PR TITLE
Combined PR: hpc-stack locations, Disable Fused Multiply Add, and Enable usage of shared pio

### LINKED ISSUE #1512: Allow shared parallelio
## Description
CMakeLists.txt hardcodes parallelio as a STATIC library. For flexibility and consistency with the way other libraries are used which allows either static or shared, can this be changed to allow for the shared version of PIO?

## Solution
Sufficiently recent versions of CMakeModules support shared PIO, at which point the PIO "STATIC" keyword in CMakeLists.txt can be removed.

### LINKED ISSUE #1619: Two HAFS forecasts crashed on WCOSS2 when Fused Multiply-Add instructions were enabled
## Description
Two HAFS forecasts crashed with the new `-DFASTER=ON` optimization options, which was tracked down to the use of fused multiply-add (FMA) instructions. Turning them off via `-no-fma` allowed the tests to pass. 

### To Reproduce:
FIXME: Someone moved the Dogwood test data, and mine is on Cactus, which I cannot access.

## Additional context

There is more information in #1618, the pull request that fixes this.

### PR DESCRIPTION
**Pull request by SamuelTrahanNOAA**
## Description
This is #1617, #1618, and #1640 combined into a single PR solely for the purposes of final testing and merge.

### Input data additions/changes
- [x] No changes are expected to input data.
- [ ] There will be new input data. 
- [ ] Input data will be updated. 

### Anticipated changes to regression tests:
- [x] Changes are expected to the following tests: hafs_regional_storm_following_1nest_atm_ocn_wav on WCOSS2 and Acorn only.

## Subcomponents involved:
- [ ] AQM
- [ ] CDEPS
- [ ] CICE
- [ ] CMEPS
- [x] CMakeModules
- [ ] FV3
- [ ] GOCART
- [ ] HYCOM
- [ ] MOM6
- [ ] NOAHMP
- [ ] WW3
- [ ] stochastic_physics
- [ ] none

### Combined with PR's (If Applicable):
This is the combination of #1617, #1618, and #1640
- Closes #1617 
- Closes #1618 
- Closes #1640 

## Commit Queue Checklist:

- [x] Link PR's from all sub-components involved
- [x] Confirm reviews completed in sub-component PR's
- [x] Add all appropriate labels to this PR.
- [x] Run full RT suite on either Hera/Cheyenne with both Intel/GNU compilers
- [ ] Add list of any failed regression tests to "Anticipated changes to regression tests" section.

## Linked PR's and Issues:

- Closes #1619 
- Closes #1512 

## Testing Day Checklist:

- [x] This PR is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR.
- [ ] Move new/updated input data on RDHPCS Hera and propagate input data changes to all supported systems.

### Testing Log (for CM's):
- RDHPCS
  - Intel
    - [x] Hera
    - [x] Orion
    - [x] Jet
    - [x] Gaea
    - [x] Cheyenne
  - GNU
    - [x] Hera
    - [x] Cheyenne
- WCOSS2
  - [x] Dogwood/Cactus
  - [x] Acorn
- CI
  - [x] Completed
- opnReqTest
  - [ ] N/A
  - [x] Log attached to comment

### CODE REVIEW

**Comment by zach1221** [comment]
@SamuelTrahanNOAA We would like to combine @natalie-perlin's PR#1617 with this PR, would you be able to pull it in?

@natalie-perlin note, on this, if possible can create a direct PR to Sam's branch?

**Comment by SamuelTrahanNOAA** [comment]
I want to make sure #1617 didn't break Rocoto support, so I'll test that now in this branch.

**Comment by SamuelTrahanNOAA** [comment]
I've combined all three PRs, and I'm testing now to ensure Rocoto support still works.

**Comment by zach1221** [comment]
Thank you, Sam!

@BrianCurtis-NOAA, PRs, #1617 #1640 and #1618 have all been combined here.

**Comment by SamuelTrahanNOAA** [comment]
I can confirm that the rt.sh Rocoto mode still works on Jet and Hera, which are the only two platforms where it is supported.

**Comment by SamuelTrahanNOAA** [comment]
@ulmononian @natalie-perlin - Could you please review this PR and make sure your changes are in here correctly?

**Comment by ulmononian** [comment]
> @ulmononian @natalie-perlin - Could you please review this PR and make sure your changes are in here correctly?

pio/cmakemodules updates look correct.

**Comment by natalie-perlin** [comment]
@jkbk2004 - 
why are totally totally different issues that have been already submitted as separate PRs are now combined into yet another PR that is very misleading - it has neither connection to original contributors and testers on various platforms, not has any basis for such additional work? This looks like a very obscured approach of what specific issues has the PR actually addresses, and it does not seem to follow accepted [GitHub Community Guidelines](https://docs.github.com/articles/github-community-guidelines).

**Comment by jkbk2004** [comment]
> @jkbk2004 - why are totally totally different issues that have been already submitted as separate PRs are now combined into yet another PR that is very misleading - it has neither connection to original contributors and testers on various platforms, not has any basis for such additional work? This looks like a very obscured approach of what specific issues has the PR actually addresses, and it does not seem to follow accepted [GitHub Community Guidelines](https://docs.github.com/articles/github-community-guidelines).

@natalie-perlin I don't get your point. The article you uploaded is only about communication in PR. Yes, we need speedy communication in PR. PRs #1618 #1640 are compile related.  Your is hpc stack modulefule update. Can you tell me why we should run full regression tests repeatedly for each PR? We have limited resource and time. Decision was made with CMs discussion. I agree new PR is not needed. But sometime mis-communication happens. @zach1221 made a note about that. I am sorry I still see no issue. I like to keep processing with this PR. BTW, thanks for your comment!

**Comment by zach1221** [comment]
Please see jenkins-ci logs attached. ORTs have passed.
[ufs-weather-model » ort-docker-pipeline » PR-1645 #1 Console [Jenkins].pdf](https://github.com/ufs-community/ufs-weather-model/files/10921187/ufs-weather-model.ort-docker-pipeline.PR-1645.1.Console.Jenkins.pdf)

File: tests/fv3_conf/fv3_qsub.IN_wcoss2 [hunk 1]
```diff
@@ -15,8 +15,8 @@ echo -n " $( date +%s )," >  job_timestamp.txt
 cd $PBS_O_WORKDIR
 
 set +x
-module use $( pwd -P )
-module load modules.fv3
+module use modulefiles
```
Comment (junwang-noaa) [inline tests/fv3_conf/fv3_qsub.IN_wcoss2]: I am curious, do we have modulefiles in the working directory?

Comment (junwang-noaa) [reply tests/fv3_conf/fv3_qsub.IN_wcoss2]: Never mind.

Comment (jkbk2004) [reply tests/fv3_conf/fv3_qsub.IN_wcoss2]: One thing I noticed from Cheyenne when admin upgraded lua last year... new lua prefers only module files in single directory. If mixed files exist in same directory, lua may complain.

**Comment by junwang-noaa** [comment]
I saw both "there will be new input data" and "input data will be updated" are checked, can someone confirm this is the case? I think we only need new baseline just because results are changed on wcoss2 because of the  hafs_regional_storm_following_1nest_atm_ocn_wav test.

**Comment by jkbk2004** [comment]
> I saw both "there will be new input data" and "input data will be updated" are checked, can someone confirm this is the case? I think we only need new baseline just because results are changed on wcoss2 because of the hafs_regional_storm_following_1nest_atm_ocn_wav test.

I think that was mishap.  I reverted the check. @SamuelTrahanNOAA was running a test yesterday, I guess. @zach1221 I am double checking on cheyenne.intel.

**Comment by SamuelTrahanNOAA** [comment]
> I think that was mishap. I reverted the check. @SamuelTrahanNOAA was running a test yesterday, I guess. @zach1221 I am double checking on cheyenne.intel.

None of the three PRs have new input data. This combined PR doesn't either.

**Comment by SamuelTrahanNOAA** [comment]
You should only need new baselines for that one test on wcoss2 and acorn. You need to run the full suite of tests on all platforms though, since the regression test system itself was changed.

**Review by DusanJovic-NOAA — APPROVED** [review]

**Comment by jkbk2004** [comment]
@zach1221 Cheyenne tests ran ok. The compile time limit issue of cpld_control_p8_faster seems resolve one more submission. We c can monitor about that issue in next PRs. I am pushing cheyenne logs.

**Comment by zach1221** [comment]
Thanks, @jkbk2004! I'll go ahead and start adding the RT labels for the other HPCs.

**Comment by BrianCurtis-NOAA** [comment]
Acorn failed to create a baseline:
```
brian.curtis@alogin01:/lfs/h2/emc/ptmp/brian.curtis/FV3_RT/rt_209021/hafs_regional_storm_following_1nest_atm_ocn_wav> more err
++ date +%s
+ echo -n ' 1678302057,'
+ cd /lfs/h2/emc/ptmp/brian.curtis/FV3_RT/rt_209021/hafs_regional_storm_following_1nest_atm_ocn_wav
+ set +x
Lmod has detected the following error: The following module(s) are unknown:
"/lfs/h2/emc/ptmp/brian.curtis/FV3_RT/rt_209021/hafs_regional_storm_following_1nest_atm_ocn_wav/modules.fv3"

Please check the spelling or version number. Also try "module spider ..."
It is also possible your cache file is out-of-date; it may help to try:
  $ module --ignore-cache load
"/lfs/h2/emc/ptmp/brian.curtis/FV3_RT/rt_209021/hafs_regional_storm_following_1nest_atm_ocn_wav/modules.fv3"

Also make sure that all modulefiles written in TCL start with the string
#%Module
```
maybe one of the modules changes are causing this issue, i'll be able to look into it in a little while.

**Comment by SamuelTrahanNOAA** [comment]
@BrianCurtis-NOAA - Try again; it should work now. When I was adding the `$PWD`, I put it on the wrong line in `fv3_qsub.IN_wcoss2`. That mistake has been corrected, and I double-checked the rest of the module logic. It looks right now.

**Comment by jkbk2004** [comment]
***Automated RT Failure Notification***
Machine: hera
Compiler: gnu
Job: RT
[RT] Repo location: /scratch1/NCEPDEV/nems/emc.nemspara/autort/pr/1267030309/20230308184516/ufs-weather-model
[RT] Error: Test cpld_control_p8 046 failed in run_test failed 
Please make changes and add the following label back: hera-gnu-RT

**Comment by zach1221** [comment]
System timeout issue on hera gnu, regarding cpld_control_p8. Will try to run manually.

**Comment by SamuelTrahanNOAA** [comment]
I reported the cpld_control_p8 problem in September of last year, and it has been going on for much longer than that:

https://github.com/ufs-community/ufs-weather-model/issues/1432

Maybe someone will fix it someday?

**Comment by DeniseWorthen** [comment]
and this one is also related https://github.com/ufs-community/ufs-weather-model/issues/1440

**Comment by zach1221** [comment]
@SamuelTrahanNOAA could you provide me with permission to push logs to this PR, please?

**Comment by SamuelTrahanNOAA** [comment]
>  Could you provide me with permission to push logs to this PR, please?

@zach1221 - you should get an invitation any second now

**Comment by jkbk2004** [comment]
***Automated RT Failure Notification***
Machine: gaea
Compiler: intel
Job: RT
[RT] Repo location: /lustre/f2/pdata/ncep/emc.nemspara/autort/pr/1267030309/20230308190009/ufs-weather-model
[RT] Error: Test cpld_control_p8_faster 021 failed in run_test failed 
Please make changes and add the following label back: gaea-intel-RT

**Comment by BrianCurtis-NOAA** [comment]
@SadeghTabas-NOAA Will the new faster RT's also need new baselines with the changes from this PR?

**Comment by sadeghitabas** [comment]
@BrianCurtis-NOAA I don't think so, I don't see any baseline changes needed to be applied for the faster tests.

**Comment by BrianCurtis-NOAA** [comment]
> @BrianCurtis-NOAA I don't think so, I don't see any baseline changes needed to be applied for the faster tests.

I think they will for WCOSS2, the -no-fma option is added to all WCOSS2 tests that use the -DFASTER compile option.

@jkbk2004 I'll need to generate new baselines for the new faster RT's on the WCOSS2 machines and run RT's again.

**Comment by jkbk2004** [comment]
> > @BrianCurtis-NOAA I don't think so, I don't see any baseline changes needed to be applied for the faster tests.
> 
> I think they will for WCOSS2, the -no-fma option is added to all WCOSS2 tests that use the -DFASTER compile option.
> 
> @jkbk2004 I'll need to generate new baselines for the new faster RT's on the WCOSS2 machines and run RT's again.

Sure!

**Comment by SamuelTrahanNOAA** [comment]
The `-DFASTER=ON` is only enabled for one test.

**Comment by BrianCurtis-NOAA** [comment]
> The `-DFASTER=ON` is only enabled for one test.

Right, we added 5 (i think) new RT's to test -DFASTER on other parts of the UFS in a previous PR. Look in rt.conf for _faster tests i.e.: https://github.com/SamuelTrahanNOAA/ufs-weather-model/blob/867add1732aeecb6b68a2580c9f37f94ba746f7d/tests/rt.conf#L126-L127

**Comment by zach1221** [comment]
Hi, @SamuelTrahanNOAA . All the testing is now complete, so we'll go ahead to merge this PR.

**Review by sadeghitabas — APPROVED** [review]

**Review by jkbk2004 — APPROVED** [review]

**Pull request merged**